scripts/map_blocks.py

```diff
@@ -0,0 +1,374 @@
+        if not (larger_shapes.crs):
+            print("Larger CRS unknown")
+            print("Larger CRS unknown")
+        if not (smaller_shapes.crs):
+            print("Smaller CRS unknown")
+            print("Smaller CRS unknown")
+    res_tuple = make_target_source_allmap(larger_shapes, smaller_shapes, larger_key, smaller_key, use_index_for_larger_key, state, year, isDemographicData)
+
+    print("Build target ==> source map")
+    final_map = {}      # {res.key: <largest of res.value>}
+    split_count = 0
+    total_blocks = 0
+    not_contained_blocks = 0
+    based_on_bb_blocks = 0
+    for key, value in res_tuple[0].items():
+        total_blocks += 1
+        if len(value) > 1:
+            split_count += 1
+            best_item = ('dummy', -2)
+            for item in value:
+                if item[1] > best_item[1]:
+                    best_item = item
+            if best_item[1] == -1:
+                based_on_bb_blocks += 1
+            final_map[key] = best_item[0]     # use [2] for precinct debugging
+        elif len(value) == 0:
+            if source_is_block_group:
+                # block not found in anything; for larger geo = block_group, assign block to block_group by id
+                bg_key = key[0:12]
+                if bg_key in res_tuple[1]:
+                    print("Block not contained in anything; using BG key: ", key)
+                    final_map[key] = bg_key
+                else:
+                    print("Block not contained in anything: ", key)
+                    final_map[key] = ''  # smaller not found in larger
+                    not_contained_blocks += 1
+            else:
+                print("Block not contained in anything: ", key)
+                final_map[key] = ''  # smaller not found in larger
+                not_contained_blocks += 1
+
+        else:
+            if value[0][1] == -1:
+                based_on_bb_blocks += 1
+            final_map[key] = value[0][0]      # use [2] for precinct debugging
+
+    source_key_set = res_tuple[1]
+    source_keys_in_map = set({})
+    for targkey, srckey in final_map.items():
+        source_keys_in_map.add(srckey)
+    print("Source keys not in map: ", len(source_key_set) - len(source_keys_in_map))
+    print("Source keys not in map: " + str(len(source_key_set) - len(source_keys_in_map)))
+
+    print("Total blocks: ", total_blocks)
+    print("Blocks not contained in anything: ", not_contained_blocks)
+    print("Blocks contained based on BB: ", based_on_bb_blocks)
+    print("Split blocks: ", split_count, "\n")
+
+    #pp = log.pretty_printer()
+    #pp.pprint(final_map)
+   
+    return final_map
+
+def make_block_map(state, large_geo_path, large_geo_key, block_geo_path, block_key, block2geo_path, year, isDemographicData, use_index_for_large_key=False, sourceIsBlkGrp=False):
+    """
+    Invokes area_contains: takes larger (precinct) geometry, smaller (block) geometry, and produces smaller ==> larger mapping (JSON)
```

Comment:
area_contains was the name for the code in data_tools, but I don't think that name means anything here.